### ISSUE TITLE
Support for self-hosted Hex repos

### ISSUE THREAD

**Issue by isaacsanders** [dependabot/dependabot-core#1286]
Released today is [MiniRepo](http://blog.plataformatec.com.br/2019/07/announcing-minirepo-a-minimal-hex-server/)

This means that more folks will need to be able to configure hex repos.

As is, the dependabot uses only the main hex repo:

https://github.com/dependabot/dependabot-core/blob/master/hex/lib/dependabot/hex/update_checker.rb#L261

Being able to configure this will allow help everyone adopting these technologies.

If I could get some guidance on adding a configuration option for this, I would be happy to provide the PR.

**Comment by asummers** [comment]
Having the URL configurable would reduce the need of [this feature](https://github.com/dependabot/dependabot-script/issues/322) in my case, though it would still be useful to others, I think.

**Comment by greysteil** [comment]
Oh awesome, I'm super up for supporting that.

Happy to give guidance on what the configuration option should look like. Our current setup for entering hex creds (for `hex.pm` organizations) looks like this:

![image](https://user-images.githubusercontent.com/1144873/62167366-f6a46000-b31a-11e9-9250-183f778ae245.png)

That looks like it's no longer fit for purpose. Any suggestions on what we should replace it with (so it works for both hex.pm and private orgs)? Indeed, can we replace it with a single construct for both or are they fundamentally different? I looked through the readme [here](https://github.com/wojtekmach/mini_repo#usage-with-mix) but all the possible details under "Usage with Rebar3" confused me...

**Comment by isaacsanders** [comment]
Add another type called "Private repo", and add the url as a configuration option.

**Comment by asummers** [comment]
I think those configs are still valid generally unless you're using an alternative Hex URL; afaik they did not remove organizations in Hex proper. I'm unsure if within a private Hex URL if you can have organizations, so it may be applicable there too. My first step would be to add a new form option for Hex URL and grey those form options out if someone sets a non-default Hex URL, I think, until the organization support is confirmed.

**Comment by greysteil** [comment]
OK, sounds like we're not sure on all the details / future developments here. It wouldn't be too painful to change later, so for now I'm happy to go with @isaacsanders's suggestion of a new config type.

What details do we need for each config? From the [docs](https://hexdocs.pm/hex/Mix.Tasks.Hex.Repo.html#content) it looks like `repo_name`, `repo_url`, `repo_auth_key` and `repo_public_key`. I would assume a credential would therefore look like:

```
{
  "type" => "hex_repo",
  "url" => "https://some.repo/etc",
  "auth_key" => "secret_key",
  "public_key" => nil
}
```

**Comment by isaacsanders** [comment]
That seems reasonable to me.

**Comment by lleger** [comment]
Just wanted to chime in and add 👍🏻 to this outstanding feature request to support Hex private repos.

You can currently use Hex organizations for private dependencies: https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#hex-organization

But you can't use private repos (which are self-hosted outside of Hex). Oban Pro, a popular job processing library in Elixir, is currently distributed in this fashion: https://hexdocs.pm/oban/pro_installation.html

When you add Oban Pro, it breaks Dependabot, which returns this error:

```
Dependency resolution failed: Unknown repository "oban", add new repositories with the `mix hex.repo add` task
```

Some of the other languages have support for this. I would think you'd need to do something like this:

```yaml
registries:
  hex-private-repo:
    type: hex-repo
    url: https://example.com/repo
    fetch_public_key: ${{secrets.HEX_KEY_FINGERPRINT}}
    auth_key: ${{secrets.HEX_AUTH_KEY}}
```

That maps to the CLI command args:

```sh
mix hex.repo add private_dep https://example.com/repo \
  --fetch-public-key $HEX_KEY_FINGERPRINT \
  --auth-key $HEX_AUTH_KEY
```

**Comment by sebasortiz-dev** [comment]
@sorentwo any workarround to get it working with oban pro meanwhile?

**Comment by sorentwo** [comment]
@sebasortiz-dev There still isn't a workaround on it.
@cam-carter Working on this currently.

**Comment by sorentwo** [comment]
A private hex repo PR is open: https://github.com/dependabot/dependabot-core/pull/5043

**Comment by jeffwidman** [comment]
👋 Have a related ask for any `Hex` / Elixir experts out there.

We  had to pin `Hex` earlier today to `1.0.1` in Dependabot due to a compiler error with the new `2.0.0` release. We obviously would rather stay on newest, but no one on the current Dependabot team is very familiar with the Hex / Elixir ecosystem. Anyone from the broader community want to take a stab at contributing a PR to fix #6009 ? At first glance, I think it's a fairly straightforward wrong-type issue, but I haven't dug deeply.

Let's keep more conversation about it over in #6009, but asking here since I know there's some Elixir experts following this issue.

**Comment by jeffwidman** [comment]
This shipped earlier today: https://github.blog/changelog/2022-11-23-dependabot-support-for-self-hosted-hex-repositories

Many thanks to @sorentwo who submitted the `dependabot-core` PR and @landongrindheim who wrote the additional code changes in our internal infra to support this! 🎉
